### PR TITLE
Improve Coprocessor Table Lookup Performance

### PR DESCRIPTION
**Pull request by breeswish**
###  What have you changed?

Introduce `incremental_get` interface at MVCC level, which will reuse the cursor when keys are given in ascending order. This can roughly improve about 30%~50% overall table lookup performance.

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

### Benchmark (Sysbench Integration)

Use Sysbench to simulate a table lookup request:

```lua
local function get_id()
  return sysbench.rand.default(1, 100000)
end

function thread_init(thread_id)
  db_connect()
end

function event(thread_id)
  local id = get_id()
  local id2 = id + 100
  rs = db_query("select * from sbtest1 where k between "..id.." AND "..id2)
end
```

Result: Table lookup 100 keys: Improve 53%

```
[ 100s ] thds: 20 tps: 242.80 qps: 242.80 (r/w/o: 242.80/0.00/0.00) lat (ms,95%): 127.81 err/s: 0.00 reconn/s: 0.00

vs

[ 100s ] thds: 20 tps: 371.40 qps: 371.40 (r/w/o: 371.40/0.00/0.00) lat (ms,95%): 92.42 err/s: 0.00 reconn/s: 0.00
```

Result: Table lookup 1000 keys: Improve 33%

```
[ 100s ] thds: 20 tps: 92.60 qps: 92.60 (r/w/o: 92.60/0.00/0.00) lat (ms,95%): 427.07 err/s: 0.00 reconn/s: 0.00

vs

[ 100s ] thds: 20 tps: 122.30 qps: 122.30 (r/w/o: 122.30/0.00/0.00) lat (ms,95%): 325.98 err/s: 0.00 reconn/s: 0.00
```

### Benchmark (Micro)

Result: Get / IncrementalGet distinct keys (common scenario): Improve 94%

```
test storage::incremental_get::bench_table_lookup_mvcc_get             ... bench:   8,636,381 ns/iter (+/- 1,348,456)
test storage::incremental_get::bench_table_lookup_mvcc_incremental_get ... bench:   4,453,101 ns/iter (+/- 984,032)
```

Result: Get / IncrementalGet continuous keys (uncommon scenario): Improve 104%

```
test storage::incremental_get::bench_table_lookup_mvcc_get             ... bench:   8,514,543 ns/iter (+/- 906,822)
test storage::incremental_get::bench_table_lookup_mvcc_incremental_get ... bench:   4,173,451 ns/iter (+/- 522,744)
```

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
good catch

**Comment by zhangjinpeng1987** [comment]
Better to have micro-bench.

**Comment by siddontang** [comment]
how about other cases? like `select count(*) from sbtest1`? 

I also notice that we use `N/A` for the memory in Cop task, can we support it later?

@breeswish

**Comment by breeswish** [comment]
@siddontang `select count(*) from sbtest1` doesn't use point get. This PR only optimize the table look up case. We may support the memory usage statistics after all data types are migrated to continuous memory.

File: src/coprocessor/dag/storage_impl.rs [hunk 1]
```diff
@@ -64,7 +64,7 @@ impl<S: Store> Storage for TiKVStorage<S> {
         let key = range.0;
         let value = self
             .store
-            .get(&Key::from_raw(&key), &mut self.cf_stats_backlog)
+            .incremental_get(&Key::from_raw(&key), &mut self.cf_stats_backlog)
```
Comment (zhangjinpeng1987) [inline src/coprocessor/dag/storage_impl.rs]: How do you guarantee it is incremental access in `fn get`?

File: src/storage/txn/store.rs [hunk 2]
```diff
@@ -158,6 +165,28 @@ impl<S: Snapshot> Store for SnapshotStore<S> {
         Ok(v)
     }
 
+    fn incremental_get(&mut self, key: &Key, statistics: &mut Statistics) -> Result<Option<Value>> {
+        if self.point_getter_cache.is_none()
+            || key.as_encoded().as_slice() < self.last_point_get_key.as_slice()
```
Comment (zhangjinpeng1987) [inline src/storage/txn/store.rs]: Can we reuse the point getter when key < last_point_get_key ？

File: src/storage/txn/store.rs [hunk 3]
```diff
@@ -250,6 +279,9 @@ impl<S: Snapshot> SnapshotStore<S> {
             start_ts,
             isolation_level,
             fill_cache,
+
+            point_getter_cache: None,
+            last_point_get_key: Vec::new(),
```
Comment (zhangjinpeng1987) [inline src/storage/txn/store.rs]: s/Vec::new()/vec![]

File: src/storage/txn/store.rs [hunk 4]
```diff
@@ -158,6 +166,28 @@ impl<S: Snapshot> Store for SnapshotStore<S> {
         Ok(v)
     }
 
+    fn incremental_get(&mut self, key: &Key) -> Result<Option<Value>> {
+        if self.point_getter_cache.is_none() {
+            // Only reuse point getter when keys are given in ascending order.
```
Comment (youjiali1995) [inline src/storage/txn/store.rs]: Why?

Comment (breeswish) [reply src/storage/txn/store.rs]: I should have deleted this comment ..

File: src/storage/mvcc/reader/point_getter.rs [hunk 5]
```diff
@@ -75,6 +75,7 @@ impl<S: Snapshot> PointGetterBuilder<S> {
         let write_cursor = CursorBuilder::new(&self.snapshot, CF_WRITE)
             .fill_cache(self.fill_cache)
             .prefix_seek(!self.multi)
+            .scan_mode(ScanMode::Mixed)
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/reader/point_getter.rs]: I think you should use ScanMode::Forward when the multi is false.

Comment (breeswish) [reply src/storage/mvcc/reader/point_getter.rs]: Nice catch. It is more strict.

Comment (zhangjinpeng1987) [reply src/storage/mvcc/reader/point_getter.rs]: Forward can prevent iterator unexpected back and forward moving.

Comment (breeswish) [reply src/storage/mvcc/reader/point_getter.rs]: Updated, PTAL again.

**Pull request merged**